nomos-libp2p/src/behaviour/nat/state_machine/tests.rs

```diff
@@ -0,0 +1,454 @@
+use std::collections::HashSet;
+
+use fixtures::*;
+use tokio::sync::mpsc::{error::TryRecvError, unbounded_channel};
+
+use super::*;
+use crate::behaviour::nat::state_machine::states::{
+    MappedPublic, Private, Public, TestIfMappedPublic, TestIfPublic, TryMapAddress,
+};
+
+impl PartialEq for Box<dyn OnEvent> {
+    fn eq(&self, other: &Self) -> bool {
+        self.box_eq(other.as_any())
+    }
+}
+
+impl PartialEq<&Self> for Box<dyn OnEvent> {
+    fn eq(&self, other: &&Self) -> bool {
+        self.box_eq(other.as_any())
+    }
+}
+
+fn all_events<'a>() -> HashSet<TestEvent<'a>> {
+    [
+        autonat_failed(),
+        autonat_ok(),
+        mapping_failed(),
+        mapping_ok(),
+        new_external_address_candidate(),
+        external_address_confirmed(),
+        default_gateway_changed(),
+        local_address_changed(),
+        other_from_swarm_event(),
+    ]
+    .into()
+}
+
+fn init_state_machine(
+    init_state: Box<dyn OnEvent>,
+    command_tx: UnboundedSender<Command>,
+) -> StateMachine {
+    StateMachine {
+        inner: Some(init_state),
+        command_tx: command_tx.into(),
+    }
+}
+
+type TestCases<'a> = Vec<(
+    // Source line number for debugging
+    u32,
+    // Factory of the initial state of the state machine
+    Box<dyn Fn() -> Box<dyn OnEvent>>,
+    // Expected transitions from the initial state
+    Vec<(
+        // Event that should cause the transition
+        TestEvent<'a>,
+        // Factory of the expected state after the transition
+        Box<dyn Fn() -> Box<dyn OnEvent>>,
+        // Expected command that should be emitted by the state machine
+        // If `None`, no command is expected
+        Option<Command>,
+    )>,
+)>;
+
+#[expect(clippy::too_many_lines, reason = "Keep the test readable")]
+#[test]
+fn test_transitions_and_emitted_commands() {
+    let expected_transitions: TestCases<'_> = vec![
+        (
+            line!(),
+            Box::new(|| Uninitialized::for_test()),
```

Comment:
I think it would be great if we could split these tests into separate functions, each testing an expected transition. In case one test case fails, we can still run the others.

Reply:
At the very least one for each starting state, but ideally one for each transition. I would have personally also split the tests for each state in a separate module and put it right after the state + transition definition, but this will do fine as well.